File: app/src/main/java/hu/kts/cmetronome/ui/workout/Countdowner.java
package hu.kts.cmetronome.ui.workout;

import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.widget.TextView;

import butterknife.BindView;
import butterknife.ButterKnife;
import hu.kts.cmetronome.R;
import hu.kts.cmetronome.Settings;
import hu.kts.cmetronome.TimeProvider;
import hu.kts.cmetronome.functional.SimpleMethod;

public class Countdowner {

    @BindView(R.id.rep_counter)
    TextView repCounterTextView;

    private SimpleMethod onFinish;
    private SimpleMethod onCancel;
    private TimeProvider timeProvider;
    Settings settings;
    private int countDownColor;
    private int normalColor;

    public Countdowner(AppCompatActivity activity, Settings settings, TimeProvider timeProvider) {
        this.onCancel = onCancel;
        this.settings = settings;
        this.timeProvider = timeProvider;
        ButterKnife.bind(this, activity);
        countDownColor = ContextCompat.getColor(activity, R.color.accent);
        normalColor = ContextCompat.getColor(activity, R.color.secondary_text);
        timeProvider.observe(activity, this::onCountDownTick);
    }

    public Countdowner setOnFinish(SimpleMethod onFinish) {
        this.onFinish = onFinish;
        return this;
    }

    public Countdowner setOnCancel(SimpleMethod onCancel) {
        this.onCancel = onCancel;
        return this;
    }

    public void onCountDownTick(long remainingSeconds) {
        repCounterTextView.setText(String.valueOf(remainingSeconds));
        if (remainingSeconds == 0) {
            repCounterTextView.setTextColor(normalColor);
            onFinish.call();
        }
    }

    public void start() {
        repCounterTextView.setTextColor(countDownColor);
        timeProvider.startDown(settings.getCountdownStartValue());
    }

    public void cancel() {
        timeProvider.stop();
        repCounterTextView.setTextColor(normalColor);
        onCancel.call();
    }
}

File: app/src/main/java/hu/kts/cmetronome/ui/workout/Countdowner.kt
package hu.kts.cmetronome.ui.workout

import android.content.Context
import androidx.core.content.ContextCompat
import hu.kts.cmetronome.R
import hu.kts.cmetronome.TimeProvider
import hu.kts.cmetronome.di.AppContext
import hu.kts.cmetronome.di.TimeProviderCountdowner
import hu.kts.cmetronome.di.WorkoutScope
import hu.kts.cmetronome.repository.WorkoutSettings
import javax.inject.Inject

@WorkoutScope
class Countdowner @Inject constructor(
        @AppContext context: Context,
        private val fragment: WorkoutFragment,
        private val settings: WorkoutSettings,
        @TimeProviderCountdowner private val timeProvider: TimeProvider
) {

    var onFinish: (() -> Unit)? = null
    var onCancel: (() -> Unit)? = null

    private val countDownColor: Int = ContextCompat.getColor(context, R.color.accent)
    private val normalColor: Int = ContextCompat.getColor(context, R.color.secondary_text)
    private val timeProviderObserver = { counter: Long -> onCountDownTick(counter) }

    init {
        timeProvider.observe(fragment.lifecycle, timeProviderObserver)
    }

    private fun onCountDownTick(remainingSeconds: Long) {
        fragment.binding.repCounterTextView.text = remainingSeconds.toString()
        if (remainingSeconds == 0L) {
            fragment.binding.repCounterTextView.setTextColor(normalColor)
            onFinish?.invoke()
        }
    }

    fun start() {
        fragment.binding.repCounterTextView.setTextColor(countDownColor)
        timeProvider.startDown(settings.countdownStartValue)
    }

    fun cancel() {
        timeProvider.stop()
        fragment.binding.repCounterTextView.setTextColor(normalColor)
        onCancel?.invoke()
    }
}
